cpp/src/rolling/window_utils.cu

```diff
@@ -0,0 +1,255 @@
+  using type = std::less<>;
+};
+template <>
+struct op_impl<which::PRECEDING, window_type::CLOSED> {
+  using type = std::less<>;
+};
+template <>
+struct op_impl<which::FOLLOWING, window_type::CLOSED> {
+  using type = std::less_equal<>;
+};
+template <which Which, window_type WindowType>
+using op_t = typename op_impl<Which, WindowType>::type;
+
+template <which Which>
+struct window_offset_impl {
+  template <typename T,
+            CUDF_ENABLE_IF(!(cudf::is_timestamp<T>() or
+                             (cudf::is_index_type<T>() and !cudf::is_unsigned<T>())))>
+  std::unique_ptr<column> operator()(column_view const&,
+                                     scalar const& length,
+                                     scalar const& offset,
+                                     window_type const window_type,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::device_async_resource_ref mr) const
+  {
+    CUDF_FAIL("Unsupported rolling window type.");
+  }
+
+  template <typename InputType, typename OffsetType, typename StrictWeakOrdering>
+  struct distance_kernel {
+    OffsetType const* length;
+    OffsetType const* offset;
+    cudf::column_device_view::const_iterator<InputType> begin;
+    cudf::column_device_view::const_iterator<InputType> end;
+    __device__ size_type operator()(size_type i)
+    {
+      if constexpr (Which == which::PRECEDING) {
+        return 1 + thrust::distance(
+                     thrust::lower_bound(
+                       thrust::seq, begin, end, *(begin + i) + *offset, StrictWeakOrdering{}),
+                     begin + i);
+      } else {
+        return thrust::distance(begin + i,
+                                thrust::lower_bound(thrust::seq,
+                                                    begin,
+                                                    end,
+                                                    *(begin + i) + (*offset + *length),
+                                                    StrictWeakOrdering{})) -
+               1;
+      }
+    }
+  };
+
+  template <typename T, typename OffsetType, typename ScalarType>
+  [[nodiscard]] std::unique_ptr<column> compute_window_bounds(
+    column_view const& input,
+    scalar const& length,
+    scalar const& offset,
+    window_type const window_type,
+    rmm::cuda_stream_view stream,
+    rmm::device_async_resource_ref mr) const
+  {
+    auto result = cudf::make_numeric_column(
+      cudf::data_type(type_to_id<size_type>()), input.size(), mask_state::UNALLOCATED, stream, mr);
+    auto const d_input_device_view = cudf::column_device_view::create(input, stream);
+    auto input_begin               = d_input_device_view->begin<T>();
+    auto input_end                 = d_input_device_view->end<T>();
+    auto const d_length            = dynamic_cast<ScalarType const&>(length).data();
+    auto const d_offset            = dynamic_cast<ScalarType const&>(offset).data();
```

Comment:
These casts are "safe" because of the combination of SFINAE and `CUDF_FAIL`s, in some sense it would be nicer if there were a better way, but I don't know what it is.

Reply:
There is the `get_scalar_device_view()` but you still have to do the cast. Honestly this looks fine to me.